File: python/316_RemoveDuplicateLetters.py
from unittest import TestCase
# https://leetcode.com/problems/remove-duplicate-letters/


class RemoveDuplicateLetters(TestCase):
    def removeDuplicateLetters(self, s):
        """
        :type s: str
        :rtype: str
        """
        done = set()
        count = {}
        for ch in s:
            count[ch] = count.get(ch, 0) + 1
        ret = []
        for ch in s:
            count[ch] -= 1
            if ch in done:
                continue
            while ret and count[ret[-1]] > 0 and ret[-1] > ch:
                done.remove(ret.pop())
            # reason: if ret[-1] cannot be remove, its previous also shouldn't remove, otherwise become bigger
            ret.append(ch)
            done.add(ch)
        return "".join(ret)

    def test1(self):
        self.assertEqual("abc", self.removeDuplicateLetters("bcabc"))

    def test2(self):
        self.assertEqual("acdb", self.removeDuplicateLetters("cbacdcbc"))

    def test3(self):
        self.assertEqual("bac", self.removeDuplicateLetters("cbac"))

File: kotlin/src/main/kotlin/316.RemoveDuplicateLetters.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/remove-duplicate-letters/

import java.util.*
import kotlin.collections.HashMap

class RemoveDuplicateLetters{
  fun removeDuplicateLetters(s: String): String {
    val done = mutableSetOf<Char>()
    val count = HashMap<Char, Int>()
    for(c in s)
      count[c] = count.getOrDefault(c, 0) + 1

    val st = Stack<Char>()
    for(c in s){
      count[c] = count[c]!! - 1
      if(c in done)
        continue
      while(st.isNotEmpty() && count[st.peek()]!! > 0 && st.peek() > c)
        done.remove(st.pop())
      st.push(c)
      done.add(c)
    }
    return st.joinToString("")
  }

  fun removeDuplicateLetters1(s: String): String {
    fun bit(ch: Char) = 1 shl (ch.toInt() - 'a'.toInt())
    val n = s.length
    if(n == 0) return ""

    val combs = IntArray(n)
    val remain = mutableSetOf<Char>()
    for(i in n-1 downTo 0){
      val ch = s[i]
      remain.add(ch)
      combs[i] = (if(i==n-1) 0 else combs[i+1]) or bit(ch)
    }
    var least = combs[0]
    var (l,r) = 0 to 0

    return buildString {
      while(least > 0){
        while(r+1 < n && (least and combs[r+1]) == least)
          r++

        val mini = (l..r).filter { s[it] in remain }.minBy { s[it] }!!    // maybe use slide-window to achieve O(1)
        val ch = s[mini]

        append(ch)
        remain.remove(ch)
        least = least xor bit(ch)
        l = mini+1
      }
    }
  }

  @Test
  fun test1(){
    assertEquals("abc", removeDuplicateLetters("bcabc"))
  }

  @Test
  fun test2(){
    assertEquals("acdb", removeDuplicateLetters("cbacdcbc"))
  }

  @Test
  fun test3(){
    assertEquals("bac", removeDuplicateLetters("cbac"))
  }


}